impl/src/test/java/io/kroxylicious/testing/kafka/testcontainers/TestcontainersKafkaClusterTest.java

```diff
@@ -0,0 +1,256 @@
+                    .isNotNull().satisfies(perImagePullPolicy -> {
+                        assertThat(perImagePullPolicy.dockerImageName().getRegistry()).isEqualTo("docker.io");
+                        assertThat(perImagePullPolicy.dockerImageName().getRepository()).isEqualTo("example/kafka-native");
+                        final ImagePullPolicy actualPullPolicy = perImagePullPolicy.pullPolicy();
+                        assertThat(actualPullPolicy).isInstanceOf(ImagePullPolicy.class); // We can't see `AlwaysPullPolicy` as it's not public
+                        assertThat(actualPullPolicy.shouldPull(perImagePullPolicy.dockerImageName())).isTrue();
+                    });
+        }
+    }
+
+    @SetEnvironmentVariable(key = TestcontainersKafkaCluster.KAFKA_IMAGE_TAG, value = "latest-kafka-3.7.0")
+    @Test
+    void shouldAllowEnvVarToControlKafkaVersion() {
+        KafkaClusterConfig config = clusterConfigBuilder.build();
+        try (TestcontainersKafkaCluster testcontainersKafkaCluster = new TestcontainersKafkaCluster(config)) {
+
+            // When
+            final TestcontainersKafkaCluster.PerImagePullPolicy kafkaImage = testcontainersKafkaCluster.getKafkaImage();
+
+            // Then
+            assertThat(kafkaImage)
+                    .isNotNull().satisfies(perImagePullPolicy -> {
+                        // this works because the classpath will be 3.9.0 or greater so we are forcing a fallback
+                        assertThat(perImagePullPolicy.dockerImageName().getVersionPart()).isEqualTo("latest-kafka-3.7.0");
+                        final ImagePullPolicy actualPullPolicy = perImagePullPolicy.pullPolicy();
+                        assertThat(actualPullPolicy).isInstanceOf(ImagePullPolicy.class); // We can't see `AlwaysPullPolicy` as it's not public
+                        assertThat(actualPullPolicy.shouldPull(perImagePullPolicy.dockerImageName())).isTrue();
+                    });
+        }
+    }
+
+    @ParameterizedTest
+    @MethodSource("fixedVersions")
+    void shouldAllowConfigToControlFixedKafkaVersion(Version version) {
+        KafkaClusterConfig config = clusterConfigBuilder.kafkaVersion(version.value()).build();
+        try (TestcontainersKafkaCluster testcontainersKafkaCluster = new TestcontainersKafkaCluster(config)) {
+
+            // When
+            final TestcontainersKafkaCluster.PerImagePullPolicy kafkaImage = testcontainersKafkaCluster.getKafkaImage();
+
+            // Then
+            assertThat(kafkaImage)
+                    .isNotNull().satisfies(perImagePullPolicy -> {
+                        assertThat(perImagePullPolicy.dockerImageName().getVersionPart()).isEqualTo("latest-kafka-" + version.value());
+                        final ImagePullPolicy actualPullPolicy = perImagePullPolicy.pullPolicy();
+                        assertThat(actualPullPolicy).isInstanceOf(ImagePullPolicy.class); // We can't see `AlwaysPullPolicy` as it's not public
```

Comment:
Hmm not convinced this should actually be this policy. Its not strictly part of this test either.

Reply:
It serves to document that you wish to assert that an always pull policy is in place, but you cannot because of the protection policy.   I'd have made the test assert the policy's class name (ugly/fragile though that is).

Reply:
Did you see my suggestion to implement our own ImagePullPolicy.  We'd implement the class to give always behaviour for tags that start latest and default otherwise.   It would be cinch to unit test too.

Reply:
I had but I hadn't realised the implications that is probably a good move. 

I think this PR then just becomes about sorting the image selection so it all happens in one testable place.